### ISSUE TITLE
TransferManager Deadlock Issue

### ISSUE THREAD

**Issue by sivamadala** [aws/aws-sdk-java#1586]
[threaddump_11023.log](https://github.com/aws/aws-sdk-java/files/1998588/threaddump_11023.log)
Hi,
I am using SDK version 1.11.321. The requirement is to upload about 2.5 tera bytes of data to s3. I am using TransferManager API with fixed number of threads. Please find the attached thread dump from the process, there are deadlocks and the process hangs in the middle of the upload.

`TransferManager xferMgr = TransferManagerBuilder.standard().withS3Client(s3Client).
					withExecutorFactory(new ExecutorFactory() {
						public ExecutorService newExecutor() {
							return Executors.newFixedThreadPool(Integer.valueOf(config.getProperty(NUMBER_OF_THREADS_PER_TASK)).);
						}
					}).withMultipartUploadThreshold(Long.valueOf(config.getProperty(MULTIPART_THRESHOLD_IN_KB))*1024).
					withMinimumUploadPartSize(Long.valueOf(config.getProperty(MINIMUM_PART_SIZE_IN_KB))*1024).build();`

I am also running 2 upload tasks in parallel using a separate thread pool executor service, each upload task creates its own TransferManager instance.

Can someone please suggest a resolution? We do not have the time to redo the project using low level API at this point in the timeline.

Appreciate your help.

**Comment by dagnir** [comment]
It looks like the threads are blocked trying to get connections from the pool. By default the pool size is 50.

Are your two `TransferManager` instances sharing the same `AmazonS3` instance?

**Comment by sivamadala** [comment]
Hi,
Each Transfermanager instance has its own s3 client, and each s3 client has its own ClientConfiguration. But it seems underneath, the http connections are pooled and shared among the threads in the entire JVM. I just bumped up the number of connections, as the default was only 10 connections. I have 100 threads per transfermanager instance, and now i bumped the max number of http connections to 200. I'll rerun the process and check if i am still running into issues. But, it seems, looking at some other posts, deadlocks is not a new issue with the TransferManager API, i hope some one has overcome this issue and can guide me through.

**Comment by dagnir** [comment]
What version of the SDK are you using? In 1.11.x the default max pool size is 50: https://github.com/aws/aws-sdk-java/blob/5497de63236dad5745d7ae35dde11a53451eabc2/aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java#L56

Do you have code that we can use to reproduce this?

**Comment by sivamadala** [comment]
You were right, the default number of connections was 50, and now i have bumped up to 200 connections, and running 2 transfer manager instances with 100 threads each. I'll update my findings shortly. Thanks.

**Comment by sivamadala** [comment]
[13211_td2.log](https://github.com/aws/aws-sdk-java/files/2006989/13211_td2.log)
So, i restarted the process with the above configuration and the process runs fine for few hours and slows considerably. So, i took another thread dump and though there are no deadlocks i see lot of threads are in blocked state. Please see the attached thread dump file as well.
`java.lang.Thread.State: BLOCKED (on object monitor)
	at org.apache.log4j.Category.callAppenders(Category.java:204)
	- waiting to lock <0xfffffffba3d57488> (a org.apache.log4j.spi.RootLogger)
	at org.apache.log4j.Category.forcedLog(Category.java:391)
	at org.apache.log4j.Category.log(Category.java:856)
	at org.apache.commons.logging.impl.Log4JLogger.debug(Log4JLogger.java:155)
	at org.apache.http.client.protocol.RequestAddCookies.process(RequestAddCookies.java:123)
	at org.apache.http.protocol.ImmutableHttpProcessor.process(ImmutableHttpProcessor.java:133)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:183)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at com.amazonaws.http.apache.client.impl.SdkHttpClient.execute(SdkHttpClient.java:72)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1238)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)`

So it seems the the root cause seems to be LOG4J, the AWS SDK is making calls to LOG appender and waiting to acquire a lock. Just wondering if there is any resolution for this?

**Comment by dagnir** [comment]
Hmm, what is your log4j config? This stacktrace shows a DEBUG log taking place.

**Comment by dagnir** [comment]
It looks like your logger is set to debug level, which would produce quite a bit of logging from the SDK. Raising it to INFO/WARN should reduce the amount of logging done.

**Comment by shorea** [comment]
@sivamadala did @dagnir suggestion work for you?

**Comment by zoewangg** [comment]
Closing due to no response. Feel free to reopen if you have any questions.
